Verify SQL MERGE access control checks

djsstarburst opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

<!-- Elaborate beyond the title of the PR as necessary to help the reviewers and maintainers.-->

This commit adds a MERGE test in BaseConnectorTest that verifies
that in order execute a MERGE, you must have appropriate access
control rights.  Specifically, if there is a source table, you must
have SELECT on the table columns referenced; if the MERGE does
inserts, you must have INSERT on the target table; if the MERGE
does deletes you must have DELETE on the target table; and if
the MERGE does updates, you must have UPDATE on the target
table.

<!-- Answer the following questions to help reviewers and maintainers
understand this PR's scope at a glance:
-->

> Is this change a fix, improvement, new feature, refactoring, or other?

It is an added test.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

None of the above.

> How would you describe this change to a non-technical end user or system administrator?

## Related issues, pull requests, and links

<!-- List any issues fixed by this PR, and provide links to other related PRs, upstream release notes, and other useful resources. For example:
* Fixes #issuenumber
* Related documentation in #issuenumber
* [Some release notes](http://usefulinfo.example.com)
-->

<!-- The following sections are filled in by the maintainer with input from the contributor:
Use :white_check_mark: or (x) to signal selection.
-->

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

dain commented on testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:
```diff
@@ -4305,6 +4311,94 @@ public void testMergeSubqueries()
+
+        // Show that without INSERT on the target table, the MERGE fails
+        withPrivilegeDenied(targetTable, INSERT_TABLE, () ->
+        assertThatThrownBy(() -> assertUpdate(baseMergeSql + insertCase))
```
I think the formatting is off here

djsstarburst replied:
Fixed.

dain reviewed (approved):
Other than the formatting issue, looks good

electrum reviewed (changes requested):
These should move to `TestAccessControl`, as discussed offline. We don't need test data since we are only testing access control checks, not the merge functionality. I created #13665 which allows testing the positive case.

djsstarburst commented:
@electrum suggested that I move this test to `TestAccessControl`.  However, that test uses the Blackhole connector, which didn't support MERGE.

I added another commit to the start of my PR that moves the MERGE access control checks up to the top of `StatementAnalysis.visitMerge`. It a good change, it's not sufficient to move the test to `TestAccessControl`.  The problem is that we don't know which if any columns of the source and target tables until `StatementAnalysis.visitTable` is called, and it make a call to get the merge rowId handle.

@electrum produced an ersatz implementation of MERGE for Blackhole, and maybe that makes sense as a followup.  But at this point, I’m mildly in favor of merging my existing PR, now with the commit that does access checks earlier.

djsstarburst commented:
> I created https://github.com/trinodb/trino/pull/13665 which allows testing the positive case.

I don't see how that PR can work, because the body of `storeMergedRows` is empty.

electrum commented:
It works because the Black Hole connector doesn't store or produce data (unless you ask it to produce nulls), so there is nothing to match or update. You can see that `BlackHolePageSink` ignores the provided page.

djsstarburst commented:
I rebased on top of tip master, which has @electrum's ersatz implementation of MERGE for the Blackhole connector.  That let me remove the test from `BaseConnectorTest` and move it to `TestAccessControl`, after I removed the final successful MERGE operation and verification, which Blackhole can't execute.

I think this now satisfies the concerns raised, and is ready to merge.

electrum commented on core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -2828,16 +2828,49 @@ protected Scope visitMerge(Merge merge, Optional<Scope> scope)
+                    .findFirst()
+                    .ifPresent(mergeCase -> accessControl.checkCanDeleteFromTable(session.toSecurityContext(), tableName));
+
+            ImmutableSet.Builder<String> allUpdateColumnNamesBuilder = ImmutableSet.builder();
```
Nit: we can use `HashSet` here with the variable named `allUpdateColumnNames` since we are just building this to pass to a function. Using an immutable builder is overkill and makes the code harder to read.

djsstarburst replied:
Changed.

electrum commented on testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:
```diff
@@ -387,12 +389,81 @@ public void testUpdateAccessControl()
+
+        assertUpdate(format("CREATE TABLE %s (nation_name VARCHAR, region_name VARCHAR)", targetTable));
+
+        assertUpdate(format("INSERT INTO %s (nation_name, region_name) VALUES ('FRANCE', 'EUROPE'), ('ALGERIA', 'AFRICA'), ('GERMANY', 'EUROPE')", targetTable), 3);
```
No need to insert data, since we aren't testing the actual merge functionality, and the blackhole connector will discard the data anyway. (so it's confusing to the reader as to why we do it)

djsstarburst replied:
Removed.

electrum commented on testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:
```diff
@@ -387,12 +389,81 @@ public void testUpdateAccessControl()
+        assertUpdate("DROP TABLE " + targetTable);
+    }
+
+    private void withPrivilegeDenied(String tableName, TestingPrivilegeType type, Runnable runnable)
```
We shouldn't need this method

djsstarburst replied:
Removed.

electrum commented on testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java:
```diff
@@ -387,12 +389,81 @@ public void testUpdateAccessControl()
+
+        // Show that without SELECT on the source table, the MERGE fails regardless of which case is included
+        for (String mergeCase : ImmutableList.of(deleteCase, updateCase, insertCase)) {
+            withPrivilegeDenied(sourceTable, SELECT_COLUMN, () ->
```
Use `assertAccessDenied`, like the other tests in this class.

djsstarburst replied:
Changed.

electrum reviewed (approved):
Looks good overall. A few minor comments. Thanks for updating this.

electrum commented:
Blackhole should be able to run the final `MERGE` using `assertAccessAllowed`. The existing one that this PR replaces runs.

djsstarburst commented:
> Blackhole should be able to run the final MERGE using assertAccessAllowed. The existing merge test using assertAccessAllowed that this PR replaces runs.

Added.

electrum commented:
Thanks!
